### ISSUE TITLE
Debug `BPMCHANGE` measure duration adjustment in Aleph-0 chart

### ISSUE THREAD

**Issue by vivaria** [vivaria/tja2fumen#22]
The desync occurs specifically at measure 34 in the chart:

![image](https://github.com/vivaria/tja2fumen/assets/76574898/0baf92d8-7e68-40b5-9acb-f474c615ee4a)

From measure 19 to measure 33, each has a duration of 240ms. In the real song, measure 34 also has a duration of ~240ms. However, in my converter, measure 34 instead has a duration of 1200ms. 

The reason for this increase from 240ms to 1200ms is due to a BPMCHANGE at measure 35:

![image](https://github.com/vivaria/tja2fumen/assets/76574898/e2cc3dea-e4e6-406b-8a7c-54d5a3507aaa)

In my converter, when a BPMCHANGE occurs, the _previous_ measure's duration is adjusted.

https://github.com/vivaria/tja2fumen/blob/b874d5c41281b95082f96d0e23385aa0d8e05c4d/src/tja2fumen/converters.py#L166-L168

This works for 99% of measures with BPMCHANGE commands, and even fixes the BPM issues that `tja2bin.exe` has.

So, I need to find out why the adjustment should _not_ be applied for Aleph-0 here, but applied for all other songs I've tested.

**Comment by joshuacwnewton** [comment]
The chart works flawlessly in TJAPlayer. The issue is with my converter.

Look at these 3 groups of measures:

```bash
#BPMCHANGE 125
#SCROLL 1
#MEASURE 15/16
4,

#SCROLL 2
#MEASURE 1/8
5,
,
,
,
,
,
,
,
,
,
,
,
,
,
,
8,

#GOGOSTART
#SCROLL 1
#BPMCHANGE 250
#MEASURE 4/4
11
```

Between the drumroll-ending note (`8,`) and the set of don notes (`11`), we have a `#BPMCHANGE` (125 -> 250). Normally, `#BPMCHANGE` will trigger this logic in my code:

https://github.com/vivaria/tja2fumen/blob/b874d5c41281b95082f96d0e23385aa0d8e05c4d/src/tja2fumen/converters.py#L166-L168

This adjustment is necessary to fix the `tja2bin.exe` desync issues due to `#BPMCHANGE` commands. This adjustment has helped fix existing songs with `#BPMCHANGE` commands, such as `ia6cho.tja`, `hol6po.tja`, etc. We know that this adjustment is necessary.

But, in this case, the adjustment causes the duration for the "`8,`" measure to be way too long (240ms -> 1200ms). In fact, when timing the official fumen (from [this youtube video](https://www.youtube.com/watch?v=xorKU8ujuyY)), it shows that the adjustment shouldn't be performed at all for this measure!!

So, maybe the BPM adjustment only needs to be applied on a conditional basis. For example, only for non-empty measures. (Since, technically, `8,` is an empty measure, because `8` isn't a note at all -- it's only purpose is to indicate where drumroll/balloon should end.)

**Comment by joshuacwnewton** [comment]
> So, maybe the BPM adjustment only needs to be applied on a conditional basis. For example, only for non-empty measures.

If I take away the adjustment for all empty measures, then this breaks the conversion for multiple other songs (`linda`, `hol6po`, `ia6cho`).

Compare:

- `ia6cho` (should apply the adjustment between measure 89 (4/5) and 90)
  ![image](https://github.com/vivaria/tja2fumen/assets/16181459/57d64ee8-061f-4258-99c3-6b756140b453)
- `hol6po` (should apply the adjustment between measure 3 (2/4) and 4): 
  ![image](https://github.com/vivaria/tja2fumen/assets/16181459/b4512794-dd81-4b22-a243-213564d720ba)
- `linda` (should apply the adjustment between measure 44 and 45):
  ![image](https://github.com/vivaria/tja2fumen/assets/16181459/4a0c0601-4718-4b84-9d2c-d1a91c44cef1)

However:

- `aleph0` (should _**not**_ apply the adjustment between measure 34 and 35):
  ![image](https://github.com/vivaria/tja2fumen/assets/16181459/9798cdde-82e5-4f25-a6f1-971b9a3c3a13)
 

----

_Maybe_ the adjustment should only be skipped for empty _full measures_ (meaning the adjustment is preserved for ia6cho and hol6po, but not linda). Linda's adjustment is only ~1.89ms, so maybe skipping it _is_ the correct thing to do? But I don't feel confident in this theory.

**Comment by joshuacwnewton** [comment]
I'm going to make the (relatively unfounded) assumption that we skip BPM adjustment for measures that contain _only_ a drumroll ending.

I have 0 reason to believe that this is correct, but I can't see any other distinction between aleph0s' chart and the other charts.

The only way I can think to investigate this is by finding more TJAs that include BPMCHANGE commands right after a drumroll ends to test the theory.

**Comment by vivaria** [comment]
The desync occurs specifically at measure 34 in the chart:

![image](https://github.com/vivaria/tja2fumen/assets/76574898/0baf92d8-7e68-40b5-9acb-f474c615ee4a)

From measure 19 to measure 33, each has a duration of 240ms. However, measure 34 has a duration of 1200ms. 

The reason for this increases due to a BPMCHANGE at measure 35:

![image](https://github.com/vivaria/tja2fumen/assets/76574898/e2cc3dea-e4e6-406b-8a7c-54d5a3507aaa)

In my converter, when a BPMCHANGE occurs, the _previous_ measure's duration is adjusted.

https://github.com/vivaria/tja2fumen/blob/b874d5c41281b95082f96d0e23385aa0d8e05c4d/src/tja2fumen/converters.py#L166-L168

This works for 99% of measures with BPMCHANGE commands, and even fixes the BPM issues that `tja2bin.exe` has.

**However**, in [the real Aleph-0 song](https://www.youtube.com/watch?v=xorKU8ujuyY), the correct duration for measure 34 is _still 240ms_. (No adjustment necessary??)

So, I need to find out why the adjustment is not applied for Aleph-0 here, but applied for all other songs I've tested.

**Comment by vivaria** [comment]
What confuses me most is that there is an almost identical measure in `linda.tja`:

![image](https://github.com/vivaria/tja2fumen/assets/76574898/1cb8eec9-a8eb-4104-a05b-3389e680d0ad)

Measure 74 is a measure that contains only the end of a drumroll portion, and measure 75 has a BPMCHANGE. 

This is exactly the same as measures 34/35 in Aleph-0:

![image](https://github.com/vivaria/tja2fumen/assets/76574898/00bf93dd-c767-4c86-af30-2879455927a2)


However, for `linda.tja`, the BPM adjustment is 100% necessary. So, why isn't it necessary for Aleph-0?

**Comment by vivaria** [comment]
**Short version:** There is no problem. Aleph-0 actually converts fine. However, I got confused into thinking there was a problem.

**Long version:** The way I managed the "BPMCHANGE adjustment" was confusing:

- Previously, I added the BPMCHANGE adjustment to the measure's "duration". However, this adjustment isn't actually  part of the duration at all! A measure's duration will stay the same (in this case, 240ms) regardless of whether the next measure has a BPMCHANGE.
- Instead, the BPMCHANGE adjustment needs to be considered only in the context of the fumenOffset for the start of the _next_ measure. This results in the exact same values, but is much easier to understand.
- Also, most critically, there is difference between the "start" fumenOffset for a measure and the "end" fumenOffset for a measure.
   - **Start**: When the notes appear on screen (to the right)
   - **End**: When the notes arrive at the judgement line and you hit the note.
   - It's possible to adjust the "start" by several hundred MS, and it can still end at the expected time, because the starting adjustment gets cancelled out by the BPM change.
   - So, to really make sure that a song is out of sync, it is important to compare the _ending_ fumenOffset times, rather than the starting fumenOffsets, since the ending times dictate when you actually hit each note.
   - If I check the ending times for Aleph-0, I can confirm that the pace of the song is actually consistent (240ms between notes), even though there was a BPMCHANGE and a big 1200ms adjustment to the start of measure 35.

To avoid this confusion in the future, I have refactored my conversion code (https://github.com/vivaria/tja2fumen/commit/57edc7aecf292df643d1848912a42a34ee1a7a00) so that it explicitly computes both the starting fumenOffset _and_ the ending fumenOffset. Also, we no longer add the adjustment to the duration, but instead add it directly to the `fumenOffsetStart` attribute of the measure.
